### PR TITLE
Boolean field interpreter transforms values as true or false, otherwise it throws exception

### PR DESCRIPTION
**Pull request by DrososTasos**
The reason behind this PR is that I thought it makes more sense to be specific on which values we convert to `true` or `false` and otherwise throw an exception. With the current implementation, if user inputs 'False' instead of 'false', it will be converted to `true`.

### CODE REVIEW

File: Importer/EventListener/InterpreterExceptionSubscriber.php [hunk 1]
```diff
@@ -50,6 +51,8 @@ private function resolveInterpreterExceptionMessage(InterpreterException $except
             return "Row is expected to contain {$exception->getExpectedSize()}, found to have {$exception->getFoundSize()} fields instead.";
         } elseif ($exception instanceof InvalidRowException) {
             return $this->buildInvalidRowExceptionMessage($exception);
+        } elseif ($exception instanceof BooleanFormatException) {
```
Comment (mkruk-u2) [inline Importer/EventListener/InterpreterExceptionSubscriber.php]: Maybe it's better to have one parent exception class for all SomethingFormatException classes and to handle it by:

```
} elseif ($exception instanceof WrongValueFormatException) {
    return $exception->getMessage();
}
```

This way we don't have to add new `elseif` every time we add new exception for invalid format.

File: Importer/Interpreter/Field/BooleanFieldInterpreter.php [hunk 2]
```diff
@@ -32,6 +47,20 @@ public function interpret(FieldConfigurationInterface $configuration, $value)
             return null;
         }
 
-        return !in_array($value, $this->falseRepresentations, true);
+        if (is_bool($value)) {
+            return $value;
+        }
+
+        $value = strtolower($value);
```
Comment (mkruk-u2) [inline Importer/Interpreter/Field/BooleanFieldInterpreter.php]: What if it's an array? Or object?
And floats will be converted to strings. Why?

File: Tests/Importer/Interpreter/Field/BooleanFieldInterpreterTest.php [hunk 3]
```diff
@@ -2,42 +2,86 @@
...
-        $this->assertTrue($interpreter->interpret($configuration, 123456789));
+        $this->setExpectedExceptionRegExp(BooleanFormatException::class);
+
+        $interpreter->interpret($configuration, $value);
+    }
+
+    /**
+     * @return array
+     */
+    public function getInvalidValuesForBooleanField()
```
Comment (mkruk-u2) [inline Tests/Importer/Interpreter/Field/BooleanFieldInterpreterTest.php]: Usually we call these methods `provide...`

File: Importer/EventListener/InterpreterExceptionSubscriber.php (hunk 1 above)
Comment (yakobe) [inline Importer/EventListener/InterpreterExceptionSubscriber.php]: Not sure it makes sense to create a class just to remove it again straight after. Maybe it's better to refactor first and then use?

**Comment by yakobe** [comment]
Do we need this, it seems a bit overkill on the matching? Could we not just `strtolower` the value before we do the comparison and then if the user writes `False` it will still match `false` and work as expected? Anything else it true.

**Comment by DrososTasos** [comment]
Well for me it does not make sense the part 'anything else is true'. How about values like 'Test', 2 or whatever as boolean. They will be imported as `True`.

Lets say user wants to import:
| Integer | Boolean | Text |
| 1           | True       | Test |

but by mistake he passes on csv:
| 1           | Test        | True |

On this case user will import row will values `1, True, 'True'`, which is wrong and most importantly, user has no idea that something went wrong.

File: Importer/EventListener/InterpreterExceptionSubscriber.php (hunk 1 above)
Comment (yakobe) [inline Importer/EventListener/InterpreterExceptionSubscriber.php]: ok. This means that we have to have a new exception for each new interpreter etc. Can we not make a generic exception to save maintenance and overhead? or do we need to know specifically what exception it is somewhere?

Comment (DrososTasos) [inline Importer/EventListener/InterpreterExceptionSubscriber.php]: Well we have a generic exception. It's the `InterpreterException`. The reason that we have other types of exceptions is because of the information we want to show to user when something goes wrong.

Comment (yakobe) [inline Importer/EventListener/InterpreterExceptionSubscriber.php]: We should not be showing exception messages to the user either way. If we had a generic message, we could build it with parameters passed from a geneic exception. This could even allow translation etc.

**Comment by yakobe** [comment]
>On this case user will import row with values 1, True, 'True', which is wrong and most importantly, user has no idea that something went wrong.

Ok, valid comment. It might help to reduce user error.

File: Importer/EventListener/InterpreterExceptionSubscriber.php [hunk 4]
```diff
@@ -42,8 +42,19 @@ public function handleException(InterpreterExceptionEvent $event)
...
     private function resolveInterpreterExceptionMessage(InterpreterException $exception)
     {
-        if ($exception instanceof DateTimeFormatException) {
-            return "Error matching date “{$exception->getValue()}” to format “{$exception->getFormat()}”.";
+        if ($exception instanceof FieldValueException) {
+            $message = "The value “{$exception->getActualValue()}” is invalid.";
+
+            if (null !== $exception->getAllowedValues()) {
+                $allowedValues = implode(', ', $exception->getAllowedValues());
+                $message .= " It should match one of the following values: “{$allowedValues}”.";
```
Comment (yakobe) [inline Importer/EventListener/InterpreterExceptionSubscriber.php]: `"The following values are valid: “{$allowedValues}”."

File: Importer/EventListener/InterpreterExceptionSubscriber.php [hunk 5]
```diff
@@ -42,8 +42,19 @@ public function handleException(InterpreterExceptionEvent $event)
      */
     private function resolveInterpreterExceptionMessage(InterpreterException $exception)
     {
-        if ($exception instanceof DateTimeFormatException) {
-            return "Error matching date “{$exception->getValue()}” to format “{$exception->getFormat()}”.";
+        if ($exception instanceof FieldValueException) {
+            $message = "The value “{$exception->getActualValue()}” is invalid.";
```
Comment (yakobe) [inline Importer/EventListener/InterpreterExceptionSubscriber.php]: do we want to translate these? or maybe add a new issue?

File: Importer/Interpreter/Exception/FieldValueException.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
+<?php
+namespace Netdudes\ImporterBundle\Importer\Interpreter\Exception;
+
+class FieldValueException extends InterpreterException
```
Comment (yakobe) [inline Importer/Interpreter/Exception/FieldValueException.php]: Perhaps `InvalidValueException` is a better name? What do you think?

Comment (DrososTasos) [inline Importer/Interpreter/Exception/FieldValueException.php]: For me the current name is better and more specific since the logic of the interpreter, is to handle each field of each row, and build a message per field when possible (Boolean and DateTime fields)

File: Importer/EventListener/InterpreterExceptionSubscriber.php (hunk 5 above)
Comment (DrososTasos) [inline Importer/EventListener/InterpreterExceptionSubscriber.php]: Well at the moment we do not translate anything in this bundle. Translating them means that we have to extract them properly for translation etc. What do you think? 

I think even if we decided to translate them I would definitely do this on a different PR

File: Importer/Interpreter/Exception/FieldValueException.php (hunk 6 above)
Comment (yakobe) [inline Importer/Interpreter/Exception/FieldValueException.php]: Yes, but the idea is to be less specific 😄 . The exception handles only invalid values. Also, "field" is a fairly broad term that does not provide that much info.

It is still under the interpreter namespace, so `InvalidValueException` is a better name thand `FieldValueException`. Unless you have another suggestion?

**Comment by yakobe** [comment]
Once this is prepared for rebase = :shipit:
